addons/binding/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/discovery/MiIoDiscovery.java

```diff
@@ -0,0 +1,302 @@
+        ThingUID uid = new ThingUID(THING_TYPE_MIIO, id);
+        // Test if the device is already known by specific ThingTypes. In that case don't use the generic thingType
+        for (ThingTypeUID typeU : NONGENERIC_THING_TYPES_UIDS) {
+            ThingUID thingUID = new ThingUID(typeU, id);
+            Thing existingThing = discoveryServiceCallback.getExistingThing(thingUID);
+            if (existingThing != null) {
+                logger.trace("Mi IO device {} already exist as thing {}: {}.", id, thingUID.toString(),
+                        existingThing.getLabel());
+                uid = thingUID;
+                break;
+            }
+            DiscoveryResult dr = discoveryServiceCallback.getExistingDiscoveryResult(thingUID);
+            if (dr != null) {
+                logger.debug("Mi IO device {} already discovered as type '{}': {}", id, dr.getThingTypeUID(),
+                        dr.getLabel());
+                uid = thingUID;
+                label = dr.getLabel();
+                break;
+            }
+        }
+        logger.debug("Discovered Mi IO Device {} ({}) at {} as {}", id, Long.parseUnsignedLong(id, 16), ip, uid);
+        if (IGNORED_TOKENS.contains(token)) {
+            logger.debug(
+                    "No token discovered for device {}. For options how to get the token, check the binding readme.",
+                    id);
+            thingDiscovered(DiscoveryResultBuilder.create(uid).withProperty(PROPERTY_HOST_IP, ip)
+                    .withProperty(PROPERTY_DID, id).withRepresentationProperty(id).withLabel(label).build());
+        } else {
+            logger.debug("Discovered token for device {}: {}", id, token);
+            thingDiscovered(DiscoveryResultBuilder.create(uid).withProperty(PROPERTY_HOST_IP, ip)
+                    .withProperty(PROPERTY_DID, id).withProperty(PROPERTY_TOKEN, token).withRepresentationProperty(id)
+                    .withLabel(label + " with token").build());
+        }
+    }
+
+    synchronized DatagramSocket getSocket() {
+        if (clientSocket != null && clientSocket.isBound()) {
+            return clientSocket;
+        }
+        try {
+            logger.debug("Getting new socket for discovery");
+            DatagramSocket clientSocket = new DatagramSocket();
+            clientSocket.setReuseAddress(true);
+            clientSocket.setBroadcast(true);
+            this.clientSocket = clientSocket;
+            return clientSocket;
+        } catch (Exception e) {
+            logger.debug("Error getting socket for discovery: {}", e.getMessage(), e);
+        }
+        return null;
+    }
+
+    private void closeSocket() {
+        if (clientSocket == null) {
+            return;
+        }
+        clientSocket.close();
+        clientSocket = null;
+    }
+
+    /**
+     * @return broadcast addresses for all interfaces
+     */
+    private Set<String> getBroadcastAddresses() {
```

Comment:
Can you use `NetUtil.getAllBroadcastAddresses()` instead?

Reply:
There were reported issues about not discovering devices if 224.0.0.50 & 224.0.0.1 were not used.
So, those I add to the regular broadcast addresses to be sure.

I'll replace part of code that cycles through the network interfaces with  NetUtil.getAllBroadcastAddresses() to simplify